File: Orchestrator/src/orchestrator/csvLogger.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { EventEmitter } from 'events';
import createCsvWriter from 'csv-writer';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * CSV Logger
 * Logs test data to CSV files with all required fields per spec
 */
export class CSVLogger extends EventEmitter {
  constructor(logsDir = path.join(__dirname, '../../data/logs')) {
    super();
    this.logsDir = logsDir;
    this.currentFile = null;
    this.csvWriter = null;
    this.writeQueue = [];
    this.isWriting = false;

    // Ensure logs directory exists
    if (!fs.existsSync(this.logsDir)) {
      fs.mkdirSync(this.logsDir, { recursive: true });
    }
  }

  /**
   * Create new CSV file for a test run
   * @param {string} runName - Test run name
   * @returns {string} File path
   */
  createLogFile(runName) {
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-');
    const filename = `uwb_test_${runName}_${timestamp}.csv`;
    const filePath = path.join(this.logsDir, filename);

    // Define CSV headers per spec
    const csvWriter = createCsvWriter.createObjectCsvWriter({
      path: filePath,
      header: [
        { id: 'timestamp', title: 'timestamp' },
        { id: 'run_name', title: 'run_name' },
        { id: 'node_id', title: 'node_id' },
        { id: 'link_distance_m', title: 'link_distance_m' },
        { id: 'env_type', title: 'env_type' },
        { id: 'channel', title: 'channel' },
        { id: 'data_rate', title: 'data_rate' },
        { id: 'tx_power_idx', title: 'tx_power_idx' },
        { id: 'pkt_rate_hz', title: 'pkt_rate_hz' },
        { id: 'jammer_label', title: 'jammer_label' },
        { id: 'total_rx', title: 'total_rx' },
        { id: 'lost_pkts', title: 'lost_pkts' },
        { id: 'crc_err', title: 'crc_err' },
        { id: 'rssi_avg_dbm', title: 'rssi_avg_dbm' },
        { id: 'snr_avg_db', title: 'snr_avg_db' },
        { id: 'preamble_q_avg', title: 'preamble_q_avg' },
        { id: 'per', title: 'per' },
      ],
    });

    this.currentFile = filePath;
    this.csvWriter = csvWriter;
    this.emit('fileCreated', filePath);

    return filePath;
  }

  /**
   * Calculate Packet Error Rate (PER)
   * PER = (lost_pkts + crc_err) / (total_rx + lost_pkts + crc_err) * 100
   * @param {number} totalRx - Total received packets
   * @param {number} lostPkts - Lost packets
   * @param {number} crcErr - CRC errors
   * @returns {number} PER percentage
   */
  calculatePER(totalRx, lostPkts, crcErr) {
    const total = totalRx + lostPkts + crcErr;
    if (total === 0) {
      return 0;
    }
    return ((lostPkts + crcErr) / total) * 100;
  }

  /**
   * Log data row
   * @param {Object} data - Data object with all required fields
   */
  async logData(data) {
    // Ensure required fields
    const row = {
      timestamp: data.timestamp || new Date().toISOString(),
      run_name: data.run_name || '',
      node_id: data.node_id || '',
      link_distance_m: data.link_distance_m || 0,
      env_type: data.env_type || '',
      channel: data.channel || 0,
      data_rate: data.data_rate || '',
      tx_power_idx: data.tx_power_idx || 0,
      pkt_rate_hz: data.pkt_rate_hz || 0,
      jammer_label: data.jammer_label || '',
      total_rx: data.total_rx || 0,
      lost_pkts: data.lost_pkts || 0,
      crc_err: data.crc_err || 0,
      rssi_avg_dbm: data.rssi_avg_dbm || 0,
      snr_avg_db: data.snr_avg_db || 0,
      preamble_q_avg: data.preamble_q_avg || 0,
      per: data.per !== undefined 
        ? data.per 
        : this.calculatePER(data.total_rx || 0, data.lost_pkts || 0, data.crc_err || 0),
    };

    // Add to write queue
    this.writeQueue.push(row);

    // Process queue
    await this.processQueue();
  }

  /**
   * Process write queue (async-safe)
   */
  async processQueue() {
    if (this.isWriting || this.writeQueue.length === 0) {
      return;
    }

    if (!this.csvWriter) {
      throw new Error('No CSV file created. Call createLogFile() first.');
    }

    this.isWriting = true;

    try {
      const rows = this.writeQueue.splice(0);
      await this.csvWriter.writeRecords(rows);
      this.emit('dataLogged', rows);
    } catch (error) {
      this.emit('error', error);
      throw error;
    } finally {
      this.isWriting = false;

      // Process remaining items in queue
      if (this.writeQueue.length > 0) {
        setImmediate(() => this.processQueue());
      }
    }
  }

  /**
   * Get current log file path
   * @returns {string|null}
   */
  getCurrentFile() {
    return this.currentFile;
  }

  /**
   * List all log files
   * @returns {Array<string>}
   */
  listLogFiles() {
    try {
      const files = fs.readdirSync(this.logsDir)
        .filter(file => file.endsWith('.csv'))
        .map(file => path.join(this.logsDir, file))
        .sort()
        .reverse(); // Most recent first
      return files;
    } catch (error) {
      this.emit('error', error);
      return [];
    }
  }

  /**
   * Read log file
   * @param {string} filePath - Path to CSV file
   * @returns {Promise<Array<Object>>}
   */
  async readLogFile(filePath) {
    try {
      const csvParser = (await import('csv-parser')).default;
      const results = [];

      return new Promise((resolve, reject) => {
        fs.createReadStream(filePath)
          .pipe(csvParser())
          .on('data', (data) => results.push(data))
          .on('end', () => resolve(results))
          .on('error', reject);
      });
    } catch (error) {
      throw error;
    }
  }
}

File: Orchestrator/python_backend/src/orchestrator/csv_logger.py
"""
CSV Logger
Logs test data to CSV files with all required fields per spec
"""

import csv
import os
from datetime import datetime
from pathlib import Path
from typing import List, Dict, Any, Optional
import asyncio


class CSVLogger:
    """CSV Logger for test data"""
    
    def __init__(self, logs_dir: Optional[str] = None):
        if logs_dir is None:
            # Default to ../../data/logs relative to this file
            logs_dir = Path(__file__).parent.parent.parent.parent / "data" / "logs"
        
        self.logs_dir = Path(logs_dir)
        self.current_file: Optional[Path] = None
        self.csv_writer: Optional[csv.DictWriter] = None
        self.file_handle: Optional[Any] = None
        self.write_queue: List[Dict[str, Any]] = []
        self.is_writing = False
        self._lock = asyncio.Lock()
        
        # Ensure logs directory exists
        self.logs_dir.mkdir(parents=True, exist_ok=True)
    
    def create_log_file(self, run_name: str) -> str:
        """
        Create new CSV file for a test run
        
        Args:
            run_name: Test run name
            
        Returns:
            File path
        """
        timestamp = datetime.now().isoformat().replace(':', '-').replace('.', '-')
        filename = f"uwb_test_{run_name}_{timestamp}.csv"
        file_path = self.logs_dir / filename
        
        # Close previous file if open
        if self.file_handle:
            self.file_handle.close()
        
        # Open new file
        self.file_handle = open(file_path, 'w', newline='')
        
        # Define CSV headers per spec
        fieldnames = [
            'timestamp',
            'run_name',
            'node_id',
            'link_distance_m',
            'env_type',
            'channel',
            'data_rate',
            'tx_power_idx',
            'pkt_rate_hz',
            'jammer_label',
            'total_rx',
            'lost_pkts',
            'crc_err',
            'rssi_avg_dbm',
            'snr_avg_db',
            'preamble_q_avg',
            'per',
        ]
        
        self.csv_writer = csv.DictWriter(self.file_handle, fieldnames=fieldnames)
        self.csv_writer.writeheader()
        self.file_handle.flush()
        
        self.current_file = file_path
        print(f"[CSVLogger] Created log file: {file_path}")
        
        return str(file_path)
    
    @staticmethod
    def calculate_per(total_rx: int, lost_pkts: int, crc_err: int) -> float:
        """
        Calculate Packet Error Rate (PER)
        PER = (lost_pkts + crc_err) / (total_rx + lost_pkts + crc_err) * 100
        """
        total = total_rx + lost_pkts + crc_err
        if total == 0:
            return 0.0
        return ((lost_pkts + crc_err) / total) * 100
    
    async def log_data(self, data: Dict[str, Any]) -> None:
        """
        Log data row
        
        Args:
            data: Data dict with required fields
        """
        # Ensure required fields
        row = {
            'timestamp': data.get('timestamp', datetime.now().isoformat()),
            'run_name': data.get('run_name', ''),
            'node_id': data.get('node_id', ''),
            'link_distance_m': data.get('link_distance_m', 0),
            'env_type': data.get('env_type', ''),
            'channel': data.get('channel', 0),
            'data_rate': data.get('data_rate', ''),
            'tx_power_idx': data.get('tx_power_idx', 0),
            'pkt_rate_hz': data.get('pkt_rate_hz', 0),
            'jammer_label': data.get('jammer_label', ''),
            'total_rx': data.get('total_rx', 0),
            'lost_pkts': data.get('lost_pkts', 0),
            'crc_err': data.get('crc_err', 0),
            'rssi_avg_dbm': data.get('rssi_avg_dbm', 0),
            'snr_avg_db': data.get('snr_avg_db', 0),
            'preamble_q_avg': data.get('preamble_q_avg', 0),
            'per': data.get('per', self.calculate_per(
                data.get('total_rx', 0),
                data.get('lost_pkts', 0),
                data.get('crc_err', 0)
            )),
        }
        
        # Add to write queue
        self.write_queue.append(row)
        
        # Process queue
        await self._process_queue()
    
    async def _process_queue(self) -> None:
        """Process write queue (async-safe)"""
        if self.is_writing or not self.write_queue:
            return
        
        if not self.csv_writer:
            raise Exception('No CSV file created. Call create_log_file() first.')
        
        async with self._lock:
            if self.is_writing or not self.write_queue:
                return
            
            self.is_writing = True
            
            try:
                rows = self.write_queue.copy()
                self.write_queue.clear()
                
                # Write rows
                for row in rows:
                    self.csv_writer.writerow(row)
                
                self.file_handle.flush()
                
            except Exception as error:
                print(f"[CSVLogger] Error writing to CSV: {error}")
                raise
            finally:
                self.is_writing = False
                
                # Process remaining items in queue
                if self.write_queue:
                    asyncio.create_task(self._process_queue())
    
    def get_current_file(self) -> Optional[str]:
        """Get current log file path"""
        return str(self.current_file) if self.current_file else None
    
    def list_log_files(self) -> List[str]:
        """List all log files"""
        try:
            files = sorted(
                [str(f) for f in self.logs_dir.glob('*.csv')],
                reverse=True  # Most recent first
            )
            return files
        except Exception as error:
            print(f"[CSVLogger] Error listing log files: {error}")
            return []
    
    async def read_log_file(self, file_path: str) -> List[Dict[str, Any]]:
        """Read log file"""
        try:
            results = []
            with open(file_path, 'r', newline='') as f:
                reader = csv.DictReader(f)
                for row in reader:
                    results.append(row)
            return results
        except Exception as error:
            raise Exception(f"Failed to read log file: {error}")
    
    def close(self) -> None:
        """Close current log file"""
        if self.file_handle:
            self.file_handle.close()
            self.file_handle = None
            self.csv_writer = None
